bitcoin/src/taproot.rs

```diff
@@ -727,6 +727,20 @@ impl_try_from!(&[TapNodeHash]);
 impl_try_from!(Vec<TapNodeHash>);
 impl_try_from!(Box<[TapNodeHash]>);
 
+macro_rules! impl_try_from_array {
+    ($($len:expr),* $(,)?) => {
+        $(
+            impl From<[TapNodeHash; $len]> for TaprootMerkleBranch {
+                fn from(a: [TapNodeHash; $len]) -> Self {
+                    Self(a.to_vec())
+                }
+            }
+        )*
+    }
+}
+// Implement for all values [1, 128] inclusive.
+impl_try_from_array!(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128);
```

Comment:
This doesn't impl 0 but `from_collection` allows it. So one of them is wrong, I don't have much time to figure out which one.

Reply:
Initialization from a zero sized array and initialization for an empty length are not exactly the same, right? Two thoughts:
- Seems like there is no reason to enable initialization from a zero sized array, if folks _really_ want to do it they can use `&[]` or `Vec::new()`.
- Currently we do not have an invariant that the inner vector is non-empty so initialization from an empty array is valid.

Reply:
IIRC Taproot rules disallow empty branch so we should have the invariant and should check it. But still didn't have time to double-check. In any case they should be consistent. I'm fine with disabling initialization from empty collections in a separate PR.

Reply:
Agreed, we should prevent it in `from_collection` (and I checked, it is not currently prevented). Also agreed that this should be a separate commit, if not a separate PR.

Reply:
Will do, added to my todo list. Thanks